amf-aml/shared/src/main/scala/amf/aml/internal/semantic/SemanticExtensionsFacade.scala

```diff
@@ -0,0 +1,114 @@
+import amf.aml.internal.render.emitters.instances.DefaultNodeMappableFinder
+import amf.aml.internal.semantic.SemanticExtensionHelper.{findAnnotationMapping, findSemanticExtension}
+import amf.core.client.scala.model.domain.{AmfObject, DomainElement}
+import amf.core.client.scala.model.domain.extensions.{CustomDomainProperty, DomainExtension}
+import amf.core.client.scala.parse.document.{ParserContext, SyamlParsedDocument, UnspecifiedReference}
+import amf.core.internal.metamodel.domain.extensions.DomainExtensionModel
+import amf.core.internal.parser.domain.Annotations
+import amf.core.internal.parser.{ParseConfiguration, Root}
+import org.mulesoft.common.core.CachedFunction
+import org.mulesoft.common.functional.MonadInstances._
+import org.yaml.model.{YDocument, YMap, YMapEntry, YNode}
+
+class SemanticExtensionsFacade private (val registry: AMLRegistry) {
+
+  def parse(extensionName: String,
+            parentTypes: Seq[String],
+            ast: YMapEntry,
+            ctx: ParserContext,
+            extensionId: String): Option[DomainExtension] = {
+    findExtensionDialect(extensionName).flatMap { dialect =>
+      findSemanticExtension(dialect, extensionName)
+        .map(findAnnotationMapping(dialect, _)) match {
+        case Some(mapping) if mapping.appliesTo(parentTypes) =>
+          Some(parseSemanticExtension(dialect, mapping, ast, ctx, extensionId))
+        case _ => None // TODO: throw warning?
+      }
+    }
+  }
+
+  def findAnnotation(extensionName: String): Option[AnnotationMapping] = {
+    findExtensionDialect(extensionName).flatMap { dialect =>
+      findSemanticExtension(dialect, extensionName).map(findAnnotationMapping(dialect, _))
+    }
+  }
+
+  def findAnnotationMappingByExtension(extensionName: String, dialect: Dialect): Option[AnnotationMapping] =
+    findSemanticExtension(dialect, extensionName).map { extension =>
+      findAnnotationMapping(dialect, extension)
+    }
+
+  def parseSemanticExtension(dialect: Dialect,
+                             mapping: AnnotationMapping,
+                             ast: YMapEntry,
+                             ctx: ParserContext,
+                             extensionId: String): DomainExtension = {
+
+    implicit val instanceCtx: DialectInstanceContext = instanceContext(dialect, ctx)
+
+    val key   = ast.key.as[String]
+    val value = ast.value
+
+    val instanceElement: DialectDomainElement = parseAnnotation(mapping, ast)
+
+    val property                   = createCustomDomainProperty(instanceElement, key, value)
+    val extension: DomainExtension = createDomainExtension(key, value, extensionId, property)
+    mergeAnnotationIntoExtension(instanceElement, extension)
+  }
+
+  private def instanceContext(dialect: Dialect, ctx: ParserContext) = {
+    new DialectInstanceContext(dialect, DefaultNodeMappableFinder(Seq(dialect)), ctx)
+  }
+
+  private def mergeAnnotationIntoExtension(instanceElement: DialectDomainElement, extension: DomainExtension) = {
```

Comment:
Not sure what was the final decision on this. Initially we spoke about maintaining the same behavior as custom domain properties. Custom domain properties emulate statically a dynamic (runtime) model. I understand this generates a dynamic model. If this is correct, you might want to check JSON-LD emission since we were hacking for this dynamic model emulation. There's an ADR for that (I believe in amf-core).

Reply:
The place where the extensions will reside is complicated as it differs between parsed model and transformed model.

In the parsed model the extension will be in the CustomDomainProperties field. The extension will have the DialectDomainElement's fields inlined in it.

In the transformed model the extension's fields will be top-level fields of it's parent. 

For now we are only worried about the JSON-LD parsing and emission of the transformed case. Surely we will have to revisit the CustomDomainProperties JSON-LD code when implementing the parsing part. 

DomainExtensions and SemanticExtensions will probably keep going down different roads.